File: lib/fluxsource/scpfluxsource.cc
#include "lib/core/globals.h"
#include "lib/data/fluxmap.h"
#include "lib/external/kryoflux.h"
#include "lib/fluxsource/fluxsource.pb.h"
#include "lib/core/utils.h"
#include "lib/fluxsource/fluxsource.h"
#include "lib/external/scp.h"
#include "lib/config/proto.h"
#include "lib/core/logger.h"
#include <fstream>

static int trackno(int strack)
{
    return strack >> 1;
}

static int headno(int strack)
{
    return strack & 1;
}

static int strackno(int track, int side)
{
    return (track << 1) | side;
}

class ScpFluxSource : public TrivialFluxSource
{
public:
    ScpFluxSource(const ScpFluxSourceProto& config): _config(config)
    {
        _if.open(_config.filename(), std::ios::in | std::ios::binary);
        if (!_if.is_open())
            error("cannot open input file '{}': {}",
                _config.filename(),
                strerror(errno));

        _if.read((char*)&_header, sizeof(_header));
        check_for_error();

        if ((_header.file_id[0] != 'S') || (_header.file_id[1] != 'C') ||
            (_header.file_id[2] != 'P'))
            error("input not a SCP file");

        _extraConfig.mutable_drive()->set_drive_type(
            (_header.flags & SCP_FLAG_96TPI) ? DRIVETYPE_80TRACK
                                             : DRIVETYPE_40TRACK);

        _resolution = 25 * (_header.resolution + 1);
        int startSide = (_header.heads == 2) ? 1 : 0;
        int endSide = (_header.heads == 1) ? 0 : 1;

        if ((_header.cell_width != 0) && (_header.cell_width != 16))
            error("currently only 16-bit cells in SCP files are supported");

        log("SCP tracks {}-{}, heads {}-{}",
            trackno(_header.start_track),
            trackno(_header.end_track),
            startSide,
            endSide);
        log("SCP sample resolution: {} ns", _resolution);
    }

public:
    std::unique_ptr<const Fluxmap> readSingleFlux(int track, int side) override
    {
        int strack = strackno(track, side);
        if (strack >= ARRAY_SIZE(_header.track))
            return std::make_unique<Fluxmap>();
        uint32_t offset = Bytes(_header.track[strack], 4).reader().read_le32();
        if (offset == 0)
            return std::make_unique<Fluxmap>();

        ScpTrackHeader trackheader;
        _if.seekg(offset, std::ios::beg);
        _if.read((char*)&trackheader, sizeof(trackheader));
        check_for_error();

        if ((trackheader.track_id[0] != 'T') ||
            (trackheader.track_id[1] != 'R') ||
            (trackheader.track_id[2] != 'K'))
            error("corrupt SCP file");

        std::vector<ScpTrackRevolution> revs(_header.revolutions);
        for (int revolution = 0; revolution < _header.revolutions; revolution++)
        {
            ScpTrackRevolution trackrev;
            _if.read((char*)&trackrev, sizeof(trackrev));
            check_for_error();
            revs[revolution] = trackrev;
        }

        auto fluxmap = std::make_unique<Fluxmap>();
        nanoseconds_t pending = 0;
        unsigned inputBytes = 0;
        for (int revolution = 0; revolution < _header.revolutions; revolution++)
        {
            if (revolution != 0)
                fluxmap->appendIndex();

            uint32_t datalength =
                Bytes(revs[revolution].length, 4).reader().read_le32();
            uint32_t dataoffset =
                Bytes(revs[revolution].offset, 4).reader().read_le32();

            Bytes data(datalength * 2);
            _if.seekg(dataoffset + offset, std::ios::beg);
            _if.read((char*)data.begin(), data.size());
            check_for_error();

            ByteReader br(data);
            for (int cell = 0; cell < datalength; cell++)
            {
                uint16_t interval = br.read_be16();
                if (interval)
                {
                    fluxmap->appendInterval(
                        (interval + pending) * _resolution / NS_PER_TICK);
                    fluxmap->appendPulse();
                    pending = 0;
                }
                else
                    pending += 0x10000;
            }

            inputBytes += datalength * 2;
        }

        return fluxmap;
    }

    void recalibrate() override {}

private:
    void check_for_error()
    {
        if (_if.fail())
            error("SCP read I/O error: {}", strerror(errno));
    }

private:
    const ScpFluxSourceProto& _config;
    std::ifstream _if;
    ScpHeader _header;
    nanoseconds_t _resolution;
};

std::unique_ptr<FluxSource> FluxSource::createScpFluxSource(
    const ScpFluxSourceProto& config)
{
    return std::unique_ptr<FluxSource>(new ScpFluxSource(config));
}

File: java/com/cowlark/fluxengine/fluxsource/ScpFluxSource.java
package com.cowlark.fluxengine.fluxsource;

import static com.cowlark.fluxengine.external.FluxEngine.NS_PER_TICK;
import static com.cowlark.fluxengine.external.Scp.SCP_FLAG_96TPI;

import com.cowlark.fluxengine.config.ConfigBuilder;
import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.core.ByteReader;
import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.core.FluxEngineException;
import com.cowlark.fluxengine.core.Logger;
import com.cowlark.fluxengine.data.CylinderHead;
import com.cowlark.fluxengine.data.Fluxmap;
import com.cowlark.fluxengine.data.Locations;
import com.cowlark.fluxengine.external.DriveType;
import com.cowlark.fluxengine.external.Scp;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * A flux source which reads an SCP flux file, ported from
 * lib/fluxsource/scpfluxsource.cc.
 */
public class ScpFluxSource extends TrivialFluxSource
{
    private final Bytes data;
    private final double resolution;
    private final int heads;
    private final int startTrack;
    private final int endTrack;
    private final int flags;
    private final int revolutions;
    private final int[] trackOffsets = new int[168];
    protected ConfigProto extraConfig;

    public ScpFluxSource(ScpFluxSourceProto config)
    {
        data = readFile(config.getFilename());

        ByteReader br = new ByteReader(data);
        byte[] fileId = br.read(3).toByteArray();
        if ((fileId[0] != 'S') || (fileId[1] != 'C') || (fileId[2] != 'P'))
            throw new FluxEngineException("input not a SCP file");

        br.read8(); /* version */
        br.read8(); /* type */
        revolutions = br.read8();
        startTrack = Scp.trackno(br.read8());
        endTrack = Scp.trackno(br.read8());
        flags = br.read8();
        int cellWidth = br.read8();
        heads = br.read8();
        int resolutionByte = br.read8();
        br.skip(4); /* checksum */

        for (int i = 0; i < 168; i++)
            trackOffsets[i] = br.readLe32();

        if ((cellWidth != 0) && (cellWidth != 16))
            throw new FluxEngineException("currently only 16-bit cells in SCP files are supported");

        resolution = 25.0 * (resolutionByte + 1);

        int startSide = (heads == 2) ? 1 : 0;
        int endSide = (heads == 1) ? 0 : 1;

        List<CylinderHead> chs = new ArrayList<>();
        for (int cylinder = startTrack; cylinder <= endTrack; cylinder++)
            for (int head = startSide; head <= endSide; head++)
                chs.add(new CylinderHead(cylinder, head));

        ConfigProto.Builder builder = ConfigProto.newBuilder();
        builder.getDriveBuilder()
                .setDriveType((flags & SCP_FLAG_96TPI) != 0 ?
                        DriveType.DRIVETYPE_80TRACK :
                        DriveType.DRIVETYPE_40TRACK);
        builder.getDriveBuilder().setTracks(Locations.convertCylinderHeadsToString(chs));
        extraConfig = builder.build();

        Logger.logf("SCP tracks %d-%d, heads %d-%d", startTrack, endTrack, startSide, endSide);
        Logger.logf("SCP sample resolution: %d ns", (int) resolution);
    }

    private static Bytes readFile(String filename)
    {
        try
        {
            return new Bytes(Files.readAllBytes(Path.of(filename)));
        } catch (IOException e)
        {
            throw new FluxEngineException(
                    "cannot open input file '" + filename + "': " + e.getMessage());
        }
    }

    @Override
    public void adjustConfig(ConfigBuilder configBuilder)
    {
        configBuilder.mergeConfig(extraConfig);
    }

    @Override
    public Fluxmap readSingleFlux(FluxReadParameters parameters)
    {
        int strack = Scp.strackno(parameters.cylinder(), parameters.head());
        if (strack >= 168)
            return new Fluxmap();
        int offset = trackOffsets[strack];
        if (offset == 0)
            return new Fluxmap();

        ByteReader br = new ByteReader(data);
        br.seek(offset);
        byte[] trackId = br.read(3).toByteArray();
        if ((trackId[0] != 'T') || (trackId[1] != 'R') || (trackId[2] != 'K'))
            throw new FluxEngineException("corrupt SCP file");
        br.read8(); /* strack */

        int[] revsLength = new int[revolutions];
        int[] revsOffset = new int[revolutions];
        for (int revolution = 0; revolution < revolutions; revolution++)
        {
            br.skip(4); /* index */
            revsLength[revolution] = br.readLe32();
            revsOffset[revolution] = br.readLe32();
        }

        Fluxmap fluxmap = new Fluxmap();
        long pending = 0;
        for (int revolution = 0; revolution < revolutions; revolution++)
        {
            if (revolution != 0)
                fluxmap.appendIndex();

            int dataLength = revsLength[revolution];
            int dataOffset = revsOffset[revolution];

            ByteReader dbr = new ByteReader(data);
            dbr.seek(dataOffset + offset);
            for (int cell = 0; cell < dataLength; cell++)
            {
                int interval = dbr.readBe16();
                if (interval != 0)
                {
                    fluxmap.appendInterval((int) ((interval + pending) * resolution / NS_PER_TICK));
                    fluxmap.appendPulse();
                    pending = 0;
                } else
                    pending += 0x10000;
            }
        }

        return fluxmap;
    }
}
